holidays/countries/singapore.py

```diff
@@ -276,24 +284,42 @@ def _add_with_observed(
             2023: (NOV, 12),
         }
         if year in dates_fixed_obs:
-            hol_date = date(year, *dates_fixed_obs[year])
-            _add_with_observed(hol_date, "Deepavali")
+            observed_dates.add(
+                self._add_holiday(  # type: ignore[arg-type]
+                    "Deepavali", *dates_fixed_obs[year]
+                )
+            )
         else:
-            hol_date = self.cnls.s_diwali_date(year)
-            _add_with_observed(hol_date, "Deepavali* (*estimated)")
+            observed_dates.add(
+                self._add_holiday(  # type: ignore[arg-type]
+                    "Deepavali* (*estimated)",
+                    self._convert_chinese_to_gre(10, 1) + td(days=-2),
+                )
+            )
 
         # Christmas Day
-        _add_with_observed(
-            date(year, DEC, 25), "Christmas Day", +2 if year == 2039 else +1
-        )
+        observed_dates.add(self._add_christmas_day("Christmas Day"))
 
         # Boxing day (up to and including 1968)
         if year <= 1968:
-            self[date(year, DEC, 26)] = "Boxing Day"
+            self._add_christmas_day_two("Boxing Day")
+
+        # Implement Section 4(2) of the Holidays Act:
+        # "if any day specified in the Schedule falls on a Sunday,
+        # the day next following not being itself a public holiday
+        # is declared a public holiday in Singapore."
+        if self.observed and year >= 1998:
+            for dt in sorted(observed_dates):
+                if not self._is_sunday(dt):
+                    continue
+                obs_date = dt + td(days=+1)
+                if obs_date in observed_dates:
+                    obs_date += td(days=+1)
+                self._add_holiday("%s (Observed)" % self[dt], obs_date)
 
         # special case (observed from previuos year)
         if self.observed and year == 2007:
-            self[date(2007, JAN, 2)] = "Hari Raya Haji (Observed)"
+            self._add_holiday("Hari Raya Haji (Observed)", JAN, 2)
```

Comment:
The if above can be simplified and moved under previous one.

// also fix the comment typo please

Reply:
Done.